pyteal/ast/router_test.py

```diff
@@ -0,0 +1,455 @@
+            pt.Router.wrap_handler(False, error_case)
+        assert error_msg in str(bug)
+
+
+def test_wrap_handler_method_call():
+    with pytest.raises(pt.TealInputError) as bug:
+        pt.Router.wrap_handler(True, not_registrable)
+    assert "method call ABIReturnSubroutine is not registrable" in str(bug)
+
+    with pytest.raises(pt.TealInputError) as bug:
+        pt.Router.wrap_handler(True, safe_clear_state_delete)
+    assert "method call should be only registering ABIReturnSubroutine" in str(bug)
+
+    ONLY_ABI_SUBROUTINE_CASES = list(
+        filter(lambda x: isinstance(x, pt.ABIReturnSubroutine), GOOD_SUBROUTINE_CASES)
+    )
+    for abi_subroutine in ONLY_ABI_SUBROUTINE_CASES:
+        wrapped: pt.Expr = pt.Router.wrap_handler(True, abi_subroutine)
+        assembled_wrapped: pt.TealBlock = assemble_helper(wrapped)
+
+        args: list[pt.abi.BaseType] = [
+            spec.new_instance()
+            for spec in typing.cast(
+                list[pt.abi.TypeSpec], abi_subroutine.subroutine.expected_arg_types
+            )
+        ]
+
+        loading: list[pt.Expr]
+
+        if abi_subroutine.subroutine.argument_count() > pt.METHOD_ARG_NUM_LIMIT:
+            sdk_last_arg = pt.abi.TupleTypeSpec(
+                *[
+                    spec
+                    for spec in typing.cast(
+                        list[pt.abi.TypeSpec],
+                        abi_subroutine.subroutine.expected_arg_types,
+                    )[pt.METHOD_ARG_NUM_LIMIT - 1 :]
+                ]
+            ).new_instance()
+            loading = [
+                arg.decode(pt.Txn.application_args[index + 1])
+                for index, arg in enumerate(args[: pt.METHOD_ARG_NUM_LIMIT - 1])
+            ]
+            loading.append(
+                sdk_last_arg.decode(pt.Txn.application_args[pt.METHOD_ARG_NUM_LIMIT])
+            )
+            for i in range(pt.METHOD_ARG_NUM_LIMIT - 1, len(args)):
+                loading.append(
+                    sdk_last_arg[i - pt.METHOD_ARG_NUM_LIMIT + 1].store_into(args[i])
+                )
+        else:
+            loading = [
+                arg.decode(pt.Txn.application_args[index + 1])
+                for index, arg in enumerate(args)
+            ]
+
+        evaluate: pt.Expr
+        if abi_subroutine.type_of() != "void":
+            output_temp = abi_subroutine.output_kwarg_info.abi_type.new_instance()
+            evaluate = pt.Seq(
+                abi_subroutine(*args).store_into(output_temp),
+                pt.abi.MethodReturn(output_temp),
+            )
+        else:
+            evaluate = abi_subroutine(*args)
+
+        actual = assemble_helper(pt.Seq(*loading, evaluate, pt.Approve()))
+        with pt.TealComponent.Context.ignoreScratchSlotEquality(), pt.TealComponent.Context.ignoreExprEquality():
+            assert actual == assembled_wrapped
+
+
+def test_contract_json_obj():
+    ONLY_ABI_SUBROUTINE_CASES = list(
```

Comment:
```suggestion
    abi_subroutines = list(
```
Recommend lower case for local variables, and all caps for global constants.